### ISSUE TITLE
Stricter typing for BaseAccordionPanel

### ISSUE THREAD

**Issue by charlie-yao** [charlie-yao/react-aria-widgets#112]
Right now, the props for `BaseAccordionPanel` extends `React.HTMLAttributes<HTMLElement>`. The issue is that we only a subset of HTML tags to be used for the component.

IDK if we can even do this because this seems like a runtime check rather than a compile time check, but it would be nice if we could change `React.HTMLAttributes<...>` based on the `tagName` prop that was passed in.

**Comment by charlie-yao** [comment]
Although... given that we're already limiting `tagName` with both compile-time AND runtime checking, extending `React.HTMLAttributes<HTMLElement>` may not be that big of a deal. Still would be cool though.

**Comment by charlie-yao** [comment]
Turns out, even though `BaseAccordionPanel` is defined as extending `React.HTMLAttributes<HTMLElement>` and not a narrower subset of HTML elements, we actually do get some compile-time checking based on the subset of HTML elements that `BaseAccordionPanel` is allowed to render.

```tsx
const { tagName, ...rest } = props;
const Component = tagName;

return (
  <Component { ...rest } />
);
```

The TS compiler knows to limit which props are passed onto `<Component>` because `Component` is based on `tagName`, and `tagName` only refers to specific HTML elements. Right now, `tagName` only allows `"div"` and `"section"`. So if you try to do something like `<Component type="submit" { ...rest } />`, TS will complain unless you expand `tagName` to include `"button"`.

An issue, at least in theory, is that that sort of type checking isn't fool proof.

**Comment by charlie-yao** [comment]
> (Note: maybe there's an issue with how our tsconfig.json is setup or something? When testing some things out in the TypeScript playground, it'll give me errors that don't show up in my dev environment. Needs further inspection...)

Tested this out a little bit.

```tsx
interface ChildProps {
  blah: string;
}

function Parent(props: any) {
  return <Child blah="hello world!" />;
}

function Child({ ...rest }: ChildProps {
  return <div { ...rest } />;
}
```

In this situation, TypeScript will throw the complaint: `Type '{ blah: string; }' has no properties in common with type 'DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>'. (tsserver 2559)`.

If you change how `<Child>` renders by adding some valid HTML attributes, e.g.:

```tsx
function Child({ ...rest }: ChildProps {
  return <div data-random="lol" { ...rest } />;
}
```

TypeScript no longer complains. If you remove the spread operator and directly pass down the `blah` prop:

```tsx
function Child({ blah }: ChildProps {
  return <div data-random="lol" blah={ blah } />;
}
```

TypeScript gives the complaint:

```
Type '{ "data-random": string; blah: string; }' is not assignable to type 'DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>'.
  Property 'blah' does not exist on type 'DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>'. (tsserver 2322)
```

This somewhat explains the differences I saw in my development environment and the TypeScript playground. `<BaseAccordionPanel>` adds other HTML attributes, whereas the setup I had in the playground only spread props onto the HTML.

I'm not sure why spreading props causes TypeScript to stop complaining because if you inspect `...rest` in the second example, it tells me:

```
(parameter) rest: {
  blah: string;
}
```

In other words, it should know that we're passing down props to a `<div>` that aren't valid HTML attributes! Strange.

Regardless, getting TypeScript to complain about which props are being spread onto `<BaseAccordionPanel>` might fall outside the scope of this project. I could probably tighten up the type checking and make it more consistent in other ways, but maybe not this one.
